v.hull: use standard C boolean type

nilason opened the pull request:
GDAL 3.3.0 introduces the use of standard C bool type. Redefining bool locally, as in `vector/v.hull/chull.c` causes failure. Since GRASS now adhere to C11 standard this PR addresses this issue by replacing the local bool with the standard C bool.

Fixes #1563

nilason commented:
If there are no objections, I will soon merge and backport this. This is needed for successful compilation with GDAL 3.3.

metzm commented:
A much easier solution could be to use `TRUE|FALSE` from https://github.com/OSGeo/grass/blob/master/include/grass/gis.h#L54, no need to `#include <stdbool.h>` or to define boolean type in `v.hull`. Many other GRASS modules use GRASS defined `TRUE|FALSE` without problems.

nilason commented:
> A much easier solution could be to use `TRUE|FALSE` from https://github.com/OSGeo/grass/blob/master/include/grass/gis.h#L54, no need to `#include <stdbool.h>` or to define boolean type in `v.hull`. Many other GRASS modules use GRASS defined `TRUE|FALSE` without problems.

I did consider using GRASS’s TRUE/FALSE https://github.com/OSGeo/grass/issues/1563#issuecomment-832251652, but in the end sided with bool, which gives incomparably clearer code. I see no reason not to use the standard bool, in fact I’d welcome its use more often in GRASS in general. Perhaps by include stdbool.h in gis.h.

metzm commented:
> I did consider using GRASS’s TRUE/FALSE [#1563 (comment)](https://github.com/OSGeo/grass/issues/1563#issuecomment-832251652), but in the end sided with bool, which gives incomparably clearer code. I see no reason not to use the standard bool, in fact I’d welcome its use more often in GRASS in general. Perhaps by include stdbool.h in gis.h.

Using GRASS's TRUE/FALSE, you don't need to `#include <stdbool.h>` and only replace true with TRUE and false with FALSE in this PR. Looks like pretty clear code in `v.hull`. What exactly is unclear in your opinion?

If you really want to `#include <stdbool.h>`, the correct place would be in `gis.h`. What would be the benefits?

nilason commented:
> > I did consider using GRASS’s TRUE/FALSE [#1563 (comment)](https://github.com/OSGeo/grass/issues/1563#issuecomment-832251652), but in the end sided with bool, which gives incomparably clearer code. I see no reason not to use the standard bool, in fact I’d welcome its use more often in GRASS in general. Perhaps by include stdbool.h in gis.h.
> 
> Using GRASS's TRUE/FALSE, you don't need to `#include <stdbool.h>` and only replace true with TRUE and false with FALSE in this PR. Looks like pretty clear code in `v.hull`. What exactly is unclear in your opinion?
> 
> If you really want to `#include <stdbool.h>`, the correct place would be in `gis.h`. What would be the benefits?

Starting with GDAL 3.3 `stdbool.h` **is** included (in ogr_api.h), which causes the issue at hand in the first place. Therefore all `bool` types declared in `chull.c` must also be redeclared as `int`'s if we go by your suggestion. Precisely this is less clear code in my opinion. (I agree v.hull is has clear code and this PR is suggesting to keep it that way, dealing with the conflict with GDAL.) Compare e.g.:

```c
bool AddOne(tVertex p);

struct tVertexStructure {
  double v[3];
  int vnum;
  tEdge duplicate;  /* pointer to incident cone edge (or NULL) */
  bool onhull;      /* T iff point on hull. */
  bool mark;        /* T iff point already processed. */
  tVertex next, prev;
};

# vs

int AddOne(tVertex p);

struct tVertexStructure {
  double v[3];
  int vnum;
  tEdge duplicate;  /* pointer to incident cone edge (or NULL) */
  int onhull;       /* T iff point on hull. */
  int mark;         /* T iff point already processed. */
  tVertex next, prev;
};
```

In my mind it's quite obvious that using `bool` over `int` is making the intent of the code much clearer. `Int` is now used in so many various ways: as ordinary integer value, return codes, bit flags and also boolean value. Often only the documentation can describe an Int's intentional use. Using bool makes these cases clear reading the code alone.

I did some testing with putting this in `gis.h`:
```c
....

/* System include files */
#include <stdio.h>
#include <stdarg.h>
#include <stdbool.h>

...

/* Define TRUE and FALSE for boolean comparisons */
#ifndef TRUE
#define TRUE true
#endif

#ifndef FALSE
#define FALSE false
#endif

...
```

then there is no need to make the include in `chull.c`, and this and all other code works without any further modification. (Note: the change to `#define TRUE true` is not really necessary as true is defined as `1` anyway and similarly for FALSE.) But with the advantage of enabling possible future modifications/additions like:
`int G_is_gisbase(const char *path)`
to 
`bool G_is_gisbase(const char *path)`
or
`static void log_error(const char *msg, int fatal)`
to
`static void log_error(const char *msg, bool fatal)`

I agree including stdbool.h in gis.h would be the better way to go. However, I can't see *any* benefits of using ints over bools for boolean values.

(see [stdbool.h specs](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdbool.h.html) for reference)

nilason commented:
I have now updated this PR, to include `stdbool.h` in `gis.h` for global availability of C99's `bool` type and values `true` and `false`. The old definitions of `TRUE` and `FALSE` remain and are fully working as before; I changed their definitions to `true` and `false` (which are the same as the `1` and `0` before), to make their equivalency unambiguous.

metzm commented:
> I have now updated this PR, to include `stdbool.h` in `gis.h` for global availability of C99's `bool` type and values `true` and `false`. The old definitions of `TRUE` and `FALSE` remain and are fully working as before; I changed their definitions to `true` and `false` (which are the same as the `1` and `0` before), to make their equivalency unambiguous.

Thanks, fine with me. We could now clean up code and use `bool` types where appropriate (there are many places where int or char are used as booleans).

nilason commented:
> Thanks, fine with me. We could now clean up code and use bool types where appropriate (there are many places where int or char are used as booleans).

@metzm Thank you, as always, for your invaluable input!

Merged and backported.
